csp/tests/impl/test_pushpulladapter.py

```diff
@@ -143,6 +143,46 @@ def graph():
         result = [out[1] for out in graph_out[0]]
         self.assertEqual(result, [1, 2, 3])
 
+    def test_adapter_engine_shutdown(self):
+        class MyPushPullAdapterImpl(PushPullInputAdapter):
+            def __init__(self, typ, data):
+                self._data = data
+                self._thread = None
+                self._running = False
+
+            def start(self, starttime, endtime):
+                self._running = True
+                self._thread = threading.Thread(target=self._run)
+                self._thread.start()
+
+            def stop(self):
+                if self._running:
+                    self._running = False
+                    self._thread.join()
+
+            def _run(self):
+                idx = 0
+                while self._running and idx < len(self._data):
+                    live, t, v = self._data[idx]
+                    idx += 1
+                    self.push_tick(live, t, v)
+                    if idx == 1:
+                        self.engine_shutdown(ValueError("Dummy exception message"))
+
+        MyPushPullAdapter = py_pushpull_adapter_def(
+            "MyPushPullAdapter", MyPushPullAdapterImpl, ts["T"], typ="T", data=list
+        )
+
+        @csp.graph
+        def graph():
+            dt = datetime.utcnow()
+            data = [(True, dt, 1), (True, dt, 2), (True, dt, 3)]
+            adapter = MyPushPullAdapter(int, data)
+            csp.print("adapter", adapter)
+
+        with self.assertRaisesRegex(Exception, "ValueError: Dummy exception message"):
+            csp.run(graph, starttime=datetime(2020, 1, 1), endtime=timedelta(seconds=1))
```

Comment:
Again, this test is missing the point a bit. The values you have are timestamped with `datetime.utcnow` but you're running in 2020 for 1-second. 

Take a look at: https://github.com/Point72/csp/blob/main/csp/tests/impl/test_pushpulladapter.py#L12.

We want to test 1) that the engine can be shutdown when replaying the initial data and 2) that the engine can be shutdown after the transition to live data (`self.flag_replay_complete()`)